File: python/soil temp/4.py
# 2024 vs 2021 air temp

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.dates as mdates
from matplotlib.dates import DateFormatter


file_path = "почва.xlsx"
# sheet_name = "темп почвы"

# xls = pd.ExcelFile(file_path)
# print("Листы в файле:", xls.sheet_names)

# try:
#     df = pd.read_excel(file_path, sheet_name=sheet_name)
#     print(f"Второй лист '{sheet_name}', первые 3 строки:")
#     print(df.head(3))
# except ValueError as e:
#     print(f"Ошибка: лист '{sheet_name}' не найден. \n{e}")



df = pd.read_excel(file_path, header=1)
print("\nКолонки, прочитанные из файла:")
print(df.columns.tolist())


df = df.rename(columns={
    df.columns[0]: 'region',
    df.columns[1]: 'station_id',
    df.columns[2]: 'station_name',
    df.columns[3]: 'date',
    df.columns[4]: 't_mean',
    df.columns[5]: 't_max',
    df.columns[6]: 't_min'
})

df = df.dropna(how='all')

df['date'] = pd.to_datetime(df['date'], format='%d.%m.%Y', errors='coerce')
df = df.dropna(subset=['date'])

for col in ['t_mean', 't_max', 't_min']:
    df[col] = pd.to_numeric(df[col], errors='coerce')

df['year'] = df['date'].dt.year
df['month'] = df['date'].dt.month

df = df[df['month'].isin([2, 3, 4])]

#первая дата таяния
def find_first_thaw(station_df):
    pos = station_df[station_df['t_mean'] > 0]
    if len(pos) == 0:
        return None
    return pos['date'].iloc[0]

results = {}

for year in sorted(df['year'].unique()):
    year_df = df[df['year'] == year].copy()

    # dT/dt
    year_df = year_df.sort_values(['station_id', 'date'])
    year_df['dT_dt'] = year_df.groupby('station_id')['t_mean'].diff()

    # первая дата таяния
    first_thaw = (year_df.groupby(['region', 'station_id'])
                           .apply(find_first_thaw)
                           .reset_index())
    first_thaw.columns = ['region', 'station_id', 'first_thaw_date']

    # ТОП скачков температуры
    top_jumps = (year_df.groupby('station_id')['dT_dt']
                 .max()
                 .sort_values(ascending=False)
                 .head(10))

    # средняя температура по месяцам
    monthly_mean = (year_df.groupby(['region', 'month'])['t_mean']
                    .mean()
                    .reset_index())

    results[year] = {
        'data': year_df,
        'first_thaw': first_thaw,
        'top_jumps': top_jumps,
        'monthly_mean': monthly_mean
    }

for year in results:
    print(f"\n=== ГОД {year} ===")

    print("\nПервые даты таяния:")
    print(results[year]['first_thaw'])

    print("\nТОП-10 резких скачков температуры:")
    print(results[year]['top_jumps'])

    print("\nСредняя температура по месяцам:")
    print(results[year]['monthly_mean'])

unique_years = sorted(df['year'].unique())
num_years = len(unique_years)

plt.figure(figsize=(14, 5 * num_years))

for i, year in enumerate(unique_years, start=1):
    year_df = results[year]['data']
    reg_daily = year_df.groupby(['region', 'date'])['t_mean'].mean().reset_index()

    plt.subplot(num_years, 1, i)

    for reg in reg_daily['region'].unique():
        sub = reg_daily[reg_daily['region'] == reg]
        plt.plot(sub['date'], sub['t_mean'], label=reg)

    plt.axhline(0, color='black', linestyle='--')
    # plt.title(f"Температура воздуха по регионам - {year}", fontsize=16, fontweight='bold')
    plt.xlabel("Дата")
    plt.ylabel("Температура, °C")
    plt.grid(True)
    plt.legend()

    date_format = DateFormatter("%d.%m.%Y")
    plt.gca().xaxis.set_major_formatter(date_format)
    plt.xticks(rotation=45)

plt.tight_layout()
# plt.savefig('air temp 2024 vs 2021')
plt.show()

File: matlab/soil temp/4.m
%% SOIL TEMPERATURE COMPARISON — MATLAB VERSION

clc; clear; close all;

file_path = "почва.xlsx";

opts = detectImportOptions(file_path, 'VariableNamingRule','preserve');
opts.DataRange = '2:100000';       
T = readtable(file_path, opts);

T.Properties.VariableNames(1:7) = { ...
    'region','station_id','station_name','date','t_mean','t_max','t_min'};

allEmpty = all(ismissing(T), 2);
T(allEmpty, :) = [];

T.date = datetime(T.date, 'InputFormat','dd.MM.yyyy', 'Format','dd.MM.yyyy');

numVars = {'t_mean','t_max','t_min'};
for i = 1:numel(numVars)
    v = T.(numVars{i});
    T.(numVars{i}) = str2double(string(v));  
end

T.year = year(T.date);
T.month = month(T.date);

T = T(ismember(T.month, [2 3 4]), :);


function d = find_first_thaw_py(date_col, t_col)
    mask = t_col > 0;
    if ~any(mask)
        d = NaT;
    else
        d = date_col(find(mask,1,'first'));
    end
end


years = unique(T.year);
results = struct;

for y = years'
    
    yearT = T(T.year == y, :);

    yearT = sortrows(yearT, {'station_id','date'});

    %% dT_dt 
    yearT.dT_dt = NaN(height(yearT),1);
    G = findgroups(yearT.station_id);

    for g = 1:max(G)
        rows = find(G == g);
        vals = yearT.t_mean(rows);
        yearT.dT_dt(rows) = [NaN; diff(vals)];  
    end

    %% first_thaw 
    [G2, reg2, st2] = findgroups(yearT.region, yearT.station_id);
    thaw = splitapply(@(d,t) find_first_thaw_py(d,t), yearT.date, yearT.t_mean, G2);

    first_thaw = table(reg2, st2, thaw, ...
        'VariableNames', {'region','station_id','first_thaw_date'});

    %% top jumps
    max_jump = splitapply(@(x) max(x, [], 'omitnan'), yearT.dT_dt, G);
    st_unique = splitapply(@(x) x(1), yearT.station_id, G);

    [sortedJ, idx] = sort(max_jump, 'descend');
    top10 = table(st_unique(idx(1:10)), sortedJ(1:10), ...
        'VariableNames', {'station_id','max_dT_dt'});

    %% monthly mean temperatures 
    [G3, reg3, mon3] = findgroups(yearT.region, yearT.month);
    m_mean = splitapply(@(x) mean(x, 'omitnan'), yearT.t_mean, G3);

    monthly_mean = table(reg3, mon3, m_mean, ...
        'VariableNames', {'region','month','t_mean_mean'});

    key = sprintf("Y%d", y);
    results.(key).data = yearT;
    results.(key).first_thaw = first_thaw;
    results.(key).top_jumps = top10;
    results.(key).monthly_mean = monthly_mean;
end


for y = years'
    key = sprintf("Y%d", y);
    fprintf("\n=== ГОД %d ===\n", y);
    disp("Первые даты таяния:");
    disp(results.(key).first_thaw);

    disp("ТОП скачков:");
    disp(results.(key).top_jumps);

    disp("Средняя температура:");
    disp(results.(key).monthly_mean);
end


figure('Position',[100 100 1400 450*numel(years)]);

for i = 1:numel(years)
    y = years(i);
    key = sprintf("Y%d", y);
    data = results.(key).data;

    [G4, reg4, date4] = findgroups(data.region, data.date);
    mean_temp = splitapply(@(x) mean(x, 'omitnan'), data.t_mean, G4);

    reg_daily = table(reg4, date4, mean_temp, ...
        'VariableNames', {'region','date','t_mean'});

    subplot(numel(years), 1, i);
    hold on;

    for r = unique(reg_daily.region)'
        mask = strcmp(reg_daily.region, r{1});
        plot(reg_daily.date(mask), reg_daily.t_mean(mask), ...
             'DisplayName', r{1}, 'LineWidth', 1.4);
    end

    yline(0,'--k','HandleVisibility','off');
    title(sprintf("%d", y), 'FontSize', 16, 'FontWeight','bold');
    xlabel("Date");
    ylabel("Temperature (°C)");
    legend show;
    grid on;
    datetick('x','dd.mm.yyyy','keeplimits');
    xtickangle(45);
end

saveas(gcf, 'soil_temp_2021vs2024.png');
